### PR TITLE
python3Packages.tensorflow: 2.4.2 -> 2.7.0

### PR DESCRIPTION
**Pull request by veprbl**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 1]
```diff
@@ -102,30 +102,92 @@ let
       wrapt
   ]);
 
-  bazel-build = buildBazelPackage {
+  rules_cc_patched = stdenv.mkDerivation {
```
Comment (ndl) [inline pkgs/development/python-modules/tensorflow/default.nix:105]: Given all the patches are darwin-specific, I'd have `darwin` somewhere in the name, e.g. `rules_cc_darwin` or smth? (same for `llvm-raw_patched` below)

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 2]
```diff
@@ -295,9 +352,9 @@ let
     fetchAttrs = {
       # cudaSupport causes fetch of ncclArchive, resulting in different hashes
       sha256 = if cudaSupport then
-        "10m6qj3kchgxfgb6qh59vc51knm9r9pkng8bf90h00dnggvv8234"
+        lib.fakeHash
```
Comment (ndl) [inline pkgs/development/python-modules/tensorflow/default.nix:355]: Why does it have to be `lib.fakeHash`? I understand that for darwin you don't need this (and won't hit this condition due to the assert anyway) but for other platforms I assume it should be a real hash?

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 3]
```diff
@@ -354,6 +411,14 @@ in buildPythonPackage {
 
   src = bazel-build.python;
 
+  postPatch = ''
```
Comment (ndl) [inline pkgs/development/python-modules/tensorflow/default.nix:414]: It's probably clear from the regexes themselves, but consider still adding a comment on why is this patching needed? (= versions relaxation due to mismatch between what Nixpkgs has and TF requires?)

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 4]
```diff
@@ -4,15 +4,15 @@
...
+, setuptools, wheel, Keras, keras-preprocessing, google-pasta
 , opt-einsum, astunparse, h5py
 , termcolor, grpcio, six, wrapt, protobuf, tensorflow-estimator_2
 , dill, flatbuffers-python, tblib, typing-extensions
 # Common deps
 , git, pybind11, which, binutils, glibcLocales, cython, perl
 # Common libraries
 , jemalloc, mpi, gast, grpc, sqlite, boringssl, jsoncpp
-, curl, snappy, flatbuffers-core, lmdb-core, icu, double-conversion, libpng, libjpeg_turbo, giflib
+, pkgs, re2, curl, snappy, flatbuffers-core, lmdb-core, icu, double-conversion, libpng, libjpeg_turbo, giflib
```
[the remark this answers is not in the document]
Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix:15]: Apparently it uses both `protobuf` and `pkgs.protobuf`.

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 2 above)
Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix:355]: I'm starting to test this on linux now, but any help will be appreciated

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 1 above)
Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix]: Done

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 3 above)
Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix:418]: Done

**Comment by tpwrules** [comment]
Would it be possible to backport this to 21.11? Keras is broken currently.

**Comment by veprbl** [comment]
@tpwrules Could you file an issue for broken Keras?

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (veprbl) [reply pkgs/development/python-modules/tensorflow/default.nix]: Nevermind. Turns out we can't use `protobuf` and `re2` – they are ignored on darwin and on linux they trigger the same issues as the ones we already have documented.

**Comment by veprbl** [comment]
@GrahamcOfBorg build libtensorflow

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 5]
```diff
@@ -28,7 +28,7 @@
 , avx2Support  ? stdenv.hostPlatform.avx2Support
 , fmaSupport   ? stdenv.hostPlatform.fmaSupport
 # Darwin deps
-, Foundation, Security
+, Foundation, Security, cctools, llvmPackages_11
 }:
 
 assert cudaSupport -> cudatoolkit != null
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tensorflow/default.nix:34]: Those packages are never null unless you override them to null. Also this hides bugs and typos.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 6]
```diff
@@ -102,30 +102,92 @@ let
       wrapt
   ]);
 
-  bazel-build = buildBazelPackage {
+  rules_cc_darwin_patched = stdenv.mkDerivation {
+    name = "rules_cc-${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tensorflow/default.nix:106]:
```suggestion
    pname = "rules_cc-${pname}
    inherit version;
```

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 7]
```diff
@@ -102,30 +102,92 @@ let
...
+
+    src = _bazel-build.deps;
+
+    prePatch = "pushd rules_cc";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tensorflow/default.nix:110]: Maybe sourceRoot ist the better option here.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 8]
```diff
@@ -102,30 +102,92 @@ let
...
+    installPhase = ''
+      runHook preInstall
+
+      mv rules_cc/ "$out"
+
+      runHook postInstall
+    '';
+  };
+  llvm-raw_darwin_patched = stdenv.mkDerivation {
+    name = "llvm-raw-${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tensorflow/default.nix:140]:
```suggestion
    pname = "llvm-raw-${pname}";
    inherit version;
```

**Comment by mikepurvis** [comment]
Not sure if it's in-scope for this particular change, but my colleague @iwanders and I have been experiencing some significant frustrations with the existing Tensorflow 2.4.2 from nixpkgs around version mismatches with protobuf and gRPC. Specifically, it seems like the existing `TF_SYSTEM_LIBS` wants to use gRPC from nixpkgs (which then propagates protobuf 3.18 from nixpkgs), whereas Tensorflow's vendored protobuf is 3.9, significantly older.

Given that the notes in that file about issues with `com_google_protobuf` are from a commit that's almost a year old, possibly there's a chance to reevaluate and try to bring some coherence to this story?

As another example, googleapis is [now packaged for Nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/googleapis-common-protos/default.nix), I believe, so it should also be possible to also avoid using the vendored copy of that one as well.

**Comment by veprbl** [comment]
> Not sure if it's in-scope for this particular change, but my colleague @iwanders and I have been experiencing some significant frustrations with the existing Tensorflow 2.4.2 from nixpkgs around version mismatches with protobuf and gRPC. Specifically, it seems like the existing `TF_SYSTEM_LIBS` wants to use gRPC from nixpkgs (which then propagates protobuf 3.18 from nixpkgs), whereas Tensorflow's vendored protobuf is 3.9, significantly older.

Could you file an issue with steps to reproduce?

> 
> Given that the notes in that file about issues with `com_google_protobuf` are from a commit that's almost a year old, possibly there's a chance to reevaluate and try to bring some coherence to this story?

That was done in the initial version of this PR, but it turned out that the issues are still there, so it had to be reverted.

> 
> As another example, googleapis is [now packaged for Nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/googleapis-common-protos/default.nix), I believe, so it should also be possible to also avoid using the vendored copy of that one as well.

Many things should be possible, but, in the case of tensorflow/bazel, many things require an effort that is disproportional to the benefits.

**Review by ndl — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**